turbinia/api/models/tasks_statistics.py

```diff
@@ -0,0 +1,204 @@
+
+    for task in task_results:
+      request_id = task.get('request_id')
+      task_type = task.get('name')
+      worker = task.get('worker_name')
+      user = task.get('requester')
+      if not task.get('run_time'):
+        log.debug(
+            f'Ignoring task {task.get("name")} in statistics because the '
+            f'run_time is not set, and it is required to calculate stats')
+        continue
+
+      # Stats for all/successful/failed tasks
+      task_stats['all_tasks'].add_task(task)
+      if task.get('successful') is True:
+        task_stats['successful_tasks'].add_task(task)
+      elif task.get('successful') is False:
+        task_stats['failed_tasks'].add_task(task)
+
+      # Stats for Tasks per Task type.
+      if task_type in task_stats['tasks_per_type']:
+        task_type_stats = task_stats['tasks_per_type'].get(task_type)
+      else:
+        task_type_stats = client.TurbiniaStats(f'Task type {task_type:s}')
+        task_stats['tasks_per_type'][task_type] = task_type_stats
+      task_type_stats.add_task(task)
+
+      # Stats per worker.
+      if worker in task_stats['tasks_per_worker']:
+        worker_stats = task_stats['tasks_per_worker'].get(worker)
+      else:
+        worker_stats = client.TurbiniaStats(f'Worker {worker:s}')
+        task_stats['tasks_per_worker'][worker] = worker_stats
+      worker_stats.add_task(task)
+
+      # Stats per submitting User.
+      if user in task_stats['tasks_per_user']:
+        user_stats = task_stats['tasks_per_user'].get(user)
+      else:
+        user_stats = client.TurbiniaStats(f'User {user:s}')
+        task_stats['tasks_per_user'][user] = user_stats
+      user_stats.add_task(task)
+
+      # Stats for the total request.  This will, for each request, calculate the
+      # start time of the earliest task and the stop time of the latest task.
+      # This will give the overall run time covering all tasks in the request.
+      task_start_time = task['last_update'] - task['run_time']
+      task_stop_time = task['last_update']
+      if request_id in requests:
+        start_time, stop_time = requests[request_id]
+        if task_start_time < start_time:
+          requests[request_id][0] = task_start_time
+        if task_stop_time > stop_time:
+          requests[request_id][1] = task_stop_time
+      else:
+        requests[request_id] = [task_start_time, task_stop_time]
+
+    # Add a fake task result for each request with our calculated times to the
+    # stats module
+    for min_time, max_time in requests.values():
+      task = {}
+      task['run_time'] = max_time - min_time
+      task_stats['requests'].add_task(task)
+
+    # Go over all stat objects and calculate them
+    for stat_obj in task_stats.values():
+      if isinstance(stat_obj, dict):
+        for inner_stat_obj in stat_obj.values():
+          inner_stat_obj.calculate_stats()
+      else:
+        stat_obj.calculate_stats()
+
+    return task_stats
+
+  def format_task_statistics(
```

Comment:
Could we add a couple tests that at least tests the high level `format_task_statistics()` / `get_workers_status()` or whatever interface makes sense to just compare the full markdown output so that a) we can see an example of what the output looks like and b) we have some kind of easy/quick test that gives us decent coverage of the other methods involved.  Similar to how some of the tasks test the report (https://github.com/google/turbinia/blob/b17f589c39355325d1e8d5fe9730358d69f773e4/turbinia/workers/analysis/wordpress_access_test.py#L31), though note that `textwrap.dedent()` is a good way to include big blocks of text inline.

Reply:
Thanks, I added both now